### PR TITLE
[WIP] Generalise `extract()` to identify named keywords from templates, and introduce `identify()` method

### PR DESCRIPTION
**Pull request by andycasey**
**This is not ready to be merged.**

This pull request attempts to generalise the `Path.extract()` method so that named keywords can be automatically identified from the templates, and so that there are fewer edge cases to deal with. It also introduces an `Path.identify()` method to identify the template(s) that match a given path.

These functions should allow the most likely data model to be identified from a path, and then the keywords of that data model to be extracted from the path.

There was a short email thread about this (with those @'d here; 2019/04/18), and I realise now that `sdss_access` is where this functionality should live. In fact, some of it already lived here as part of `Path.extract()`.

The cases that are *not* covered by this PR are those where the template gives an explicit Python string format for the expected match (e.g., 1 below), or when the templates are not explicit enough to distinguish a match as it is (e.g., 2 below). For example:

1. The `"fpc"` template:

```python 
"$PHOTO_REDUX/{rerun}/{run}/objcs/{camcol}/fpC-{run:0>6}-{filter}{camcol}-{field:0>4}.fit"
```

To match examples like:
```python
ebosswork/eboss/photo/redux/1/45/objcs/3/fpC-000045-g3-0123.fit
```

Where the expected result is:
```python
{'rerun': '1', 'field': '0123', 'filter': 'g', 'camcol': '3', 'run': '000045'}
```

2. The `"galaxy"` template:
```python
"$BOSS_LSS_REDUX/galaxy_{dr}{version}_{sample}_{ns}.fits.gz"
```

To match examples like
```python
"ebosswork/eboss/lss/galaxy_DR12v1.0_1_n.fits.gz"
```

Where the expected result is:
```python
{'sample': '1', 'dr': 'DR12', 'version': 'v1.0', 'ns': 'n'}
```

The existing code deals explicitly with these edge cases. A more generalised way would be advantageous, but in order to do that we would need to let the keywords in the templates be able to give a regular expression match (e.g., `{dr:DR\d+}` and `{version:v[\d\.]+}`). This is because the same flexibility is not available through Python string formatting (e.g., we can't even specify what the maximum length of the string should be). 

The only alternative is to write in edge cases for all situations where the general regular expression named search will fail.

@havok2063, @joelbrownstein, @nbmurphy : Do you think we should change the template formatting to be able to optionally include a regular expression? Or should we code in every edge case for existing templates and be more conscientious of this issue when defining future template paths?

### CODE REVIEW

**Comment by havok2063** [comment]
@andycasey this looks like a good start.  I feel we want to keep the template formatting as simple as possible.  While including regex formatting would help us capture these cases and be more explicit, I can see the path definitions getting out of hand quite quickly with complex regex.  I think regex turns a lot of people off and would make the templates less readable.  I doubt we'll have a lot of edges cases.  When building the original function, I went through and only identified a handful of existing edge cases; mostly cases with neighboring brackets, `{a}{b}`.    I would advocate for training people to be more conscientious when defining their path templates.  I think we should definitely create some **path definition guidelines** outlining some rules and best practices to minimize the number of edge cases we might get.  We can add this to the package documentation.    I have some additional comments that I'll add in the files.

**Comment by havok2063** [comment]
I wrote some tests for `Path.extract` in `tests/test_paths.py`.  Can you make sure all the tests for `extract` still pass, and add any new edge cases you've encountered to the existing test?  And add some tests for your new `identify` method?

File: python/sdss_access/path/path.py [hunk 1]
```diff
@@ -195,64 +246,18 @@ def extract(self, name, example):
         assert name in self.lookup_names(), '{0} must be a valid template name'.format(name)
         template = self.templates[name]
 
-        # expand the environment variable
+        # expand the environment variables
         template = os.path.expandvars(template)
 
-        # handle special functions; perform a drop in replacement
-        if re.match('%spectrodir', template):
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: This section of code, lines 201-209 I put in to handle the special function defined in paths using the `%function_name` syntax.  These are functions that get called and replace the `%function_name` with some new path definition subset, which may include the use of keywords.  I wrote this to ensure that `extract` properly returned any keyword arguments used in those functions.  For example, the `plateLines` path defintion.  Does your new `_extract` method handle these cases as well?

File: python/sdss_access/path/path.py [hunk 2]
```diff
@@ -716,3 +721,41 @@ def definitiondir(self, filetype, **kwargs):
...
+
+    # escape the envvar $, any dots, and forward slashes
+    subtemp = template.replace('$', '\\$') \
+                      .replace('.', '\\.') \
+                      .replace('/', '\/')
+
+    # define named search pattern.
+    named_search = re.sub('{(\w+)}', '(?P<\\1>[\\w\\d+-_]+)', subtemp)
+
+    # fix duplicates by prepending '_' to them.
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: Since repeated keywords in the template specify the same variable, we will never have two of the same keywords in a given template have different values.  It's probably safe to keep only one of the instances, and drop the `_xxx` duplicate.

File: python/sdss_access/path/path.py [hunk 3]
```diff
@@ -716,3 +721,41 @@ def definitiondir(self, filetype, **kwargs):
 
 class AccessError(Exception):
     pass
+
+
+def _extract(template, example):
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: Do you have any preferences on this being a separate standalone function or a `static method` attached to the `BasePath` class?

File: python/sdss_access/path/path.py [hunk 4]
```diff
@@ -716,3 +721,41 @@ def definitiondir(self, filetype, **kwargs):
...
+
+
+    match = re.compile(named_search).search(example)
+    if match is None:
+        return dict()
+
+    path_dict = match.groupdict()
+
+    # remove duplicates.
+    duplicate_keys = []
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: oh it looks like you drop them here....

File: python/sdss_access/path/path.py [hunk 5]
```diff
@@ -195,64 +246,18 @@ def extract(self, name, example):
...
-                value = part[1]
-                if re.findall('{(.*?)}', part[0]):
-                    # get the key name inside the brackets
-                    keys = re.findall('{(.*?)}', part[0])
-                    # remove the type : designation
-                    keys = [k.split(':')[0] for k in keys]
-                    # handle double bracket edge cases; remove this when better solution found
-                    if len(keys) > 1:
-                        if keys[0] == 'dr':
-                            # for {dr}{version}
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: this chunk of code is my attempt at manually handling the `{}{}` edge cases.  It may be easier to manually catch edge cases and add each new one we discover to the test suite.  There can't be **that** many edge cases, right??

File: python/sdss_access/path/path.py (hunk 3 above)
Comment (joelbrownstein) [reply python/sdss_access/path/path.py]: +1 BasePath

File: python/sdss_access/path/path.py (hunk 5 above)
Comment (joelbrownstein) [reply python/sdss_access/path/path.py]: Very few edge cases, would be my expectation also.

File: python/sdss_access/path/path.py [hunk 6]
```diff
@@ -716,3 +721,41 @@ def definitiondir(self, filetype, **kwargs):
...
+
+def _extract(template, example):
+
+    # escape the envvar $, any dots, and forward slashes
+    subtemp = template.replace('$', '\\$') \
+                      .replace('.', '\\.') \
+                      .replace('/', '\/')
+
+    # define named search pattern.
+    named_search = re.sub('{(\w+)}', '(?P<\\1>[\\w\\d+-_]+)', subtemp)
```
Comment (havok2063) [inline python/sdss_access/path/path.py]: Is this the correct syntax for a named capture group with a capture number, `'(?P<\\1>[\\w\\d+-_]+)'`?  Trying to run this function as is gives me the error `error: bad escape \w at position 8` here.  If I make each string in the `re.sub` raw, then the substitution works but the function fails with `error: bad character in group name '\\1' at position 6` on line `match = re.compile(named_search).search(example)`.  If I replace the `\\1` capture number with `name` , then the function returns but calls all variables the `name` reference.   I'm not familiar enough with regex named capture groups and capture group numbers to suss this out.  I'll keep playing with it though.

Comment (joelbrownstein) [reply python/sdss_access/path/path.py]: like to have a package to help do `regex-alchemy` for python

**Comment by andycasey** [comment]
Just a comment to say thanks for the review @joelbrownstein and @havok2063, and I will update this PR in the coming weeks.

I think I tend to agree about regex being too complicated for the typical muggle, and so my updates will reflect that.

I also agree about some kind of path definition guidelines, which overlaps with other recent discussions on the telecons.
